api/utils/sshutils/ssh.go

```diff
@@ -250,3 +260,263 @@ func KeysEqual(ak, bk ssh.PublicKey) bool {
+		return nil, trace.Wrap(fmt.Errorf("pub does not appear to be an ssh-rsa public key: %T", err))
+	}
+	teleportPublicKey = strings.TrimSuffix(strings.TrimPrefix(teleportPublicKey, "ssh-rsa "), "\n")
+	// we use the trimmed public key from here on out
+	teleportPublicKeyBytes := []byte(teleportPublicKey)
+
+	// the PPK file contains an anti-tampering MAC which is made up of various values which appear in the file.
+	// copied from Section C.3 of https://the.earth.li/~sgtatham/putty/0.76/htmldoc/AppendixC.html#ppk:
+	// hex-mac-data is a hexadecimal-encoded value, 64 digits long (i.e. 32 bytes), generated using the HMAC-SHA-256 algorithm with the following binary data as input:
+	// string: the algorithm-name header field.
+	// string: the encryption-type header field.
+	// string: the key-comment-string header field.
+	// string: the binary public key data, as decoded from the base64 lines after the 'Public-Lines' header.
+	// string: the plaintext of the binary private key data, as decoded from the base64 lines after the 'Private-Lines' header.
+
+	// these values are also used in the MAC generation, so we declare them as variables
+	keyType := "ssh-rsa"
+	encryptionType := "none"
+	// TODO(gus): is there a way we can get the proxy/user pair name in here?
+	// annoyingly this has to be done at generation time because the comment is part of the MAC
+	fileComment := "teleport-generated-ppk"
+
+	// create a buffer to hold the elements needed to generate the MAC
+	macInput := new(bytes.Buffer)
+
+	// string: the algorithm-name header field.
+	macKeyType, err := getRFC4251String([]byte(keyType))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.LittleEndian, macKeyType)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macKeyType to buffer: %T", err))
+	}
+
+	// string: the encryption-type header field.
+	macEncryptionType, err := getRFC4251String([]byte(encryptionType))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.BigEndian, macEncryptionType)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macEncryptionType to buffer: %T", err))
+	}
+
+	// string: the key-comment-string header field.
+	macComment, err := getRFC4251String([]byte(fileComment))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.BigEndian, macComment)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macComment to buffer: %T", err))
+	}
+
+	// base64-decode the Teleport public key, as we need its binary representation to generate the MAC
+	teleportPublicKeyDecoded := make([]byte, base64.StdEncoding.EncodedLen(len(teleportPublicKeyBytes)))
+	teleportPublicKeyByteCount, err := base64.StdEncoding.Decode(teleportPublicKeyDecoded, teleportPublicKeyBytes)
+	publicKeyData := make([]byte, teleportPublicKeyByteCount)
+	copy(publicKeyData, teleportPublicKeyDecoded)
+	if err != nil {
```

Comment:
I would move this error check so it's immediately after `err` is assigned - the further away the check gets the greater the chance for bugs.

Reply:
The docs say:

```
func (enc *Encoding) Decode(dst, src []byte) (n int, err error)
    Decode decodes src using the encoding enc. It writes at most
    DecodedLen(len(src)) bytes to dst and returns the number of bytes written.
    If src contains invalid base64 data, it will return the number of bytes
    successfully written and CorruptInputError. New line characters (\r and \n)
    are ignored.
```

So I think something like this is fine:

```
n, err := base64.StdEncoding.Decode(decoded, publicKeyBytes)
if err != nil { ... }
decoded = decoded[:n]
```